core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java

```diff
@@ -0,0 +1,185 @@
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.BundleEvent;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceReference;
+import org.osgi.util.tracker.BundleTracker;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.yaml.snakeyaml.Yaml;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+
+public class CatalogBomScanner {
+
+    private static final Logger LOG = LoggerFactory.getLogger(CatalogBomScanner.class);
+    private static final String CATALOG_BOM_URL = "catalog.bom";
+    private static final String BROOKLYN_CATALOG = "brooklyn.catalog";
+    private static final String BROOKLYN_LIBRARIES = "brooklyn.libraries";
+
+    private CatalogPopulator catalogTracker;
+
+    public void bind(ServiceReference<ManagementContext> managementContext) throws Exception {
+        LOG.debug("Binding management context");
+        catalogTracker = new CatalogPopulator(managementContext);
+    }
+
+    public void unbind(ServiceReference<ManagementContext> managementContext) throws Exception {
+        LOG.debug("Unbinding management context");
+        if (null != catalogTracker) {
+            CatalogPopulator temp = catalogTracker;
+            catalogTracker = null;
+            temp.close();
+        }
+    }
+
+    private String[] bundleIds(Bundle bundle) {
+        return new String[] {
+            String.valueOf(bundle.getBundleId()), String.valueOf(bundle.getState()), bundle.getSymbolicName()
+        };
+    }
+
+
+    public class CatalogPopulator extends BundleTracker<Long> {
+
+        private ServiceReference<ManagementContext> mgmtContextReference;
+        private ManagementContext managementContext;
+
+        public CatalogPopulator(ServiceReference<ManagementContext> serviceReference) {
+            super(serviceReference.getBundle().getBundleContext(), Bundle.ACTIVE, null);
+            this.mgmtContextReference = serviceReference;
+            open();
+        }
+
+        @Override
+        public void open() {
+            managementContext = mgmtContextReference.getBundle().getBundleContext().getService(mgmtContextReference);
+            super.open();
+        }
+
+        @Override
+        public void close() {
+            super.close();
+            managementContext = null;
+            mgmtContextReference.getBundle().getBundleContext().ungetService(mgmtContextReference);
+        }
+
+        public ManagementContext getManagementContext() {
+            return managementContext;
+        }
+
+        @Override
+        public Long addingBundle(Bundle bundle, BundleEvent bundleEvent) {
+
+            final BundleContext bundleContext = FrameworkUtil.getBundle(CatalogBomScanner.class).getBundleContext();
+            if (bundleContext == null) {
+                LOG.info("Bundle context not yet established for bundle {} {} {}", bundleIds(bundle));
```

Comment:
A short mention of the implications is also interesting (i.e. ignoring catalog.bom from xxx bundle).
The message is not correct - the context is not set for CatalogBomScanner's bundle, not for `bundle`.